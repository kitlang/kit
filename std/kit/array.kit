/**
 * Array is a heap-allocated, length-prefixed, fixed-length collection type.
 */
abstract Array[T]: Slice[T] {
    public static function new(allocator: Box[Allocator], length: Size): Self {
        var data: Ptr[T] = allocator.calloc(sizeof T * length);
        return struct Self {
            length,
            data,
        };
    }

    public function blit(other: Ptr[Array[T]], start: Size, length: Size): Void {
        memcpy(&(other.data[start]) as Ptr[Void], this.data as Ptr[Void], length * sizeof T);
    }

    public function copy(allocator: Box[Allocator]): Array[T] using implicit allocator {
        var a = Self.new(this.length);
        this.blit(a, 0, this.length);
        return a;
    }

    public function free(allocator: Box[Allocator]) {
        allocator.free(this.data);
    }
}
