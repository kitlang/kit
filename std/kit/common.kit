#[builtin] abstract CArray[T, $N = 0]: Ptr[T] {
    rules {
        ($this.length) => N as Size;

        (for $ident in $this {$e}) => for __i in 0 ... N {
            var $ident: T = $this[__i];
            {$e}
        }
    }
}
#[builtin] abstract Box[T];
