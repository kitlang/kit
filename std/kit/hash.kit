 /**
 * Hashable trait must be implemented for any key in the hashmap,
 * as the hash function is used to determine it's place in the map.
 */
trait Hashable {
    function hash(): Int;
}

enum HashCellState {
    Empty;
    Deleted;
    Occupied;
}

implement Hashable for Bool {
    function hash(): Int {
        return if this then 1 else 0;
    }
}

implement Hashable for CString {
    function hash(): Int {
        var hash = 5381;
        for char in this {
            hash = (hash << 5) + hash + char;
        }
        return hash;
    }
}

implement Hashable for Int {
    function hash(): Int {
        return this;
    }
}

implement Hashable for Char {
    function hash(): Int {
        return this;
    }
}

implement Hashable for String {
    function hash(): Int {
        var hash = 5381;
        for char in this {
            hash = (hash << 5) + hash + char;
        }
        return hash;
    }
}