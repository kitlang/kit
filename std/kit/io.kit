/**
 * An abstract reader.
 */
trait Reader {
    public function readBytes(buf: Ptr[Void], bytes: Size): Size;
    public function close(): Void;
}

/**
 * An abstract writer.
 */
trait Writer {
    public function writeBytes(buf: Ptr[Void], bytes: Size): Size;

    public function write(v: Box[Writable]): Size {
        return this.writeBytes(v.ptr(), v.size());
    }

    public function flush(): Void {}
    public function close(): Void {}
}

trait Writable {
    public function ptr(): Ptr[Void];
    public function size(): Size;
}

implement Writable for CString {
    public function ptr(): Ptr[Void] {
        return this;
    }

    public function size(): Size {
        return this.length + 1;
    }
}
