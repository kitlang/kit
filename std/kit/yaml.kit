include "yaml.h" => "yaml";

enum Yaml {
    YamlNull;
    YamlScalar(s: CString);
    YamlSequence(values: Array[Yaml]);
    YamlMap(pairs: Array[(CString, Yaml)]);

    public static function parse(allocator: Box[Allocator], s: CString): Option[Yaml] using implicit allocator {
        // FIXME: do this more efficienty by counting nodes and using a LinearAllocator
        var parser: yaml_parser_t;
        yaml_parser_initialize(parser);

        parser.yaml_parser_set_input_string(s as Ptr[Void], s.length);

        var doc: yaml_document_t;
        if yaml_parser_load(parser, doc) == 0 {
            return None;
        }

        var result = Self.getNode(doc, yaml_document_get_root_node(doc));

        yaml_document_delete(doc);
        yaml_parser_delete(parser);

        return Some(result);
    }

    static function getNode(allocator: Box[Allocator], doc: Ptr[yaml_document_t], node: Ptr[yaml_node_t]): Yaml {
        match node.type {
            YAML_SCALAR_NODE => {
                // TODO
                var val = (node.data.scalar.value as Ptr[Void] as CString).copy();
                return YamlScalar(val);
            }
            YAML_SEQUENCE_NODE => {
                var a: Array[Yaml] = Array.new(node.data.sequence.items.top - node.data.sequence.items.start);
                var i = 0;
                var p: Ptr[yaml_node_item_t] = node.data.sequence.items.start;
                while p < node.data.sequence.items.top {
                    var valueNode = yaml_document_get_node(doc, p);
                    var value = Self.getNode(doc, valueNode);
                    a[i] = value;
                    ++p;
                    ++i;
                }
                return YamlSequence(a);
            }
            YAML_MAPPING_NODE => {
                var a: Array[(CString, Yaml)] = Array.new(node.data.mapping.pairs.top - node.data.mapping.pairs.start);
                var i = 0;
                var p: Ptr[yaml_node_pair_t] = node.data.mapping.pairs.start;
                while p < node.data.mapping.pairs.top {
                    var key = yaml_document_get_node(doc, p.key);
                    var keyStr = (key.data.scalar.value as Ptr[Void] as CString).copy();
                    var valueNode = yaml_document_get_node(doc, p.value);
                    var value = Self.getNode(doc, valueNode);
                    a[i] = (keyStr, value);
                    ++p;
                    ++i;
                }
                return YamlMap(a);
            }
            default => {
                return YamlNull;
            }
        }
    }

    public function free(allocator: Box[Allocator]) {
        match this {
            YamlSequence(values) => {
                for value in values {
                    value.free();
                }
                values.free();
            }
            YamlMap(pairs) => {
                for pair in pairs {
                    // FIXME
                    // allocator.free(pair[0]);
                    pair[1].free();
                }
                pairs.free();
            }
        }
    }
}
