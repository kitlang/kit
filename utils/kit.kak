# http://www.kitlang.org
# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾

# Detection
# ‾‾‾‾‾‾‾‾‾

hook global BufCreate .*\.(kit) %{
    set-option buffer filetype kit
}

# Highlighters
# ‾‾‾‾‾‾‾‾‾‾‾‾

add-highlighter shared/kit regions
add-highlighter shared/kit/code default-region group
add-highlighter shared/kit/string  region '"' (?<!\\)(\\\\)*"         fill string
add-highlighter shared/kit/comment region '//' '$'                     fill comment

add-highlighter shared/kit/code/ regex %{\b(true|false|None|Some)\b} 0:value
add-highlighter shared/kit/code/ regex \b(struct|enum|union|abstract|function|implement|specialize|trait|typedef|var|private|public|static|const|inline|using|return|throw|this|for|as|in|if|then|else|do|while|match|default|unsafe|rule|rules|implicit|sizeof|import|print|printf)\b 0:keyword
add-highlighter shared/kit/code/ regex \b(CArray|CString|Char|Ptr|Size|Short|Long|Byte|Float|Double|UInt8|UInt16|UInt32|UInt64|Int|Int8|Int16|Int32|Int64|Float32|Float64|)\b 0:type

# Initialization
# ‾‾‾‾‾‾‾‾‾‾‾‾‾‾

hook -group kit-highlight global WinSetOption filetype=kit %{ add-highlighter window/kit ref kit }
hook -group kit-highlight global WinSetOption filetype=(?!kit).* %{ remove-highlighter window/kit }
