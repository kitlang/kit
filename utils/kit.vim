" Vim syntax file
"
" Language: Kit
" Last Change: 2018-11-08

if exists("b:current_syntax")
  finish
endif
let b:current_syntax = "kit"


syn region kitCommentOneLine start="//" end="$"
syn region kitCommentMultiLine start="/\*" end="\*/"

syn keyword kitConstant null empty this Self Some None Cons Empty
syn region kitStringOneLine start=/"/ skip=/\\"/ end=/"/ oneline
syn region kitStringMultiLine start=/"""/ end=/"""/
syn match kitCharacter /c'.'/
" syn match kitNumber ...
syn keyword kitBoolean true false
" syn match kitFloat ...

" syn match kitIdentifier ...
" syn match kitFunction ...

" syn keyword kitStatement ...
syn keyword kitConditional if then else match
syn keyword kitRepeat for do while
syn keyword kitLabel default
syn keyword kitOperator sizeof
syn keyword kitKeyword function trait implement return as break defer
syn keyword kitException throw

" syn keyword kitPreProc ...
syn keyword kitInclude include import
" syn keyword kitDefine ...
syn keyword kitMacro rules using rule implicit
" syn keyword kitPreCondit ...

syn keyword kitType Int8 Int16 Short Int32 Int64 Long
syn keyword kitType Uint8 Byte Uint16 Unt32 Uint64
syn keyword kitType Float32 Float Float64 Double
syn keyword kitType Void Bool Char Int Size Ptr
syn keyword kitType Numeric Integral NumericMixed
syn keyword kitType List CArray CString Box
syn keyword kitType Iterable Iterator Option
syn keyword kitType Allocator LinearAllocator StackAllocator
syn keyword kitType Path
syn keyword kitStorageClass public static
syn keyword kitStorageClass var const
" syn keyword kitStorageClass register volatile
syn keyword kitStructure struct union enum abstract
syn keyword kitTypedef typedef specialize

" syn keyword kitSpecial
" syn keyword kitSpecialChar
" syn keyword kitTag
" syn keyword kitDelimiter
" syn keyword kitSpecialComment
" syn keyword kitDebug

" syn keyword kitUnderlined
" syn keyword kitIgnore
" syn keyword kitError
" syn keyword kitTodo


hi def link kitCommentOneLine Comment
hi def link kitCommentMultiLine Comment

hi def link kitConstant Constant
hi def link kitStringOneLine String
hi def link kitStringMultiLine String
hi def link kitCharacter Character
" hi def link kitNumber Number
hi def link kitBoolean Boolean
" hi def link kitFloat Float

" hi def link kitIdentifier Identifier
" hi def link kitFunction Function

" hi def link kitStatement Statement
hi def link kitConditional Conditional
hi def link kitRepeat Repeat
hi def link kitLabel Label
hi def link kitOperator Operator
hi def link kitKeyword Keyword
hi def link kitException Exception

" hi def link kitPreProc Preproc
hi def link kitInclude Include
" hi def link kitDefine Define
hi def link kitMacro Macro
" hi def link kitPreCondit PreCondit

hi def link kitType Type
hi def link kitStorageClass StorageClass
hi def link kitStructure Structure
hi def link kitTypedef Typedef

" hi def link kitSpecial Special
" hi def link kitSpecialChar SpecialChar
" hi def link kitTag Tag
" hi def link kitDelimiter Delimiter
" hi def link kitSpecialComment SpecialComment
" hi def link kitDebug Debug

" hi def link kitUnderlined Underlined
" hi def link kitIgnore Ignore
" hi def link kitError Error
" hi def link kitTodo Todo
