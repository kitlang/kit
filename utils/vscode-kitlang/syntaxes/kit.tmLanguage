<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>fileTypes</key>
    <array>
      <string>kit</string>
    </array>
    <key>name</key>
    <string>Kit</string>
    <key>scopeName</key>
    <string>source.kit</string>
    <key>uuid</key>
    <string>4fc8848b-af63-4ac5-850f-8e6b00a45004</string>
    <key>patterns</key>
    <array>
      <dict>
        <key>match</key>
        <string>\b(import|include|function|implement|specialize|var|using)\b</string>
        <key>name</key>
        <string>keyword.control.kit</string>
      </dict>
      <dict>
        <key>match</key>
        <string>\b(private|public|static|const|inline)\b</string>
        <key>name</key>
        <string>storage.modifier.kit</string>
      </dict>
      <dict>
        <key>match</key>
        <string>\b(return|throw|for|as|in|if|then|else|do|while|match|default|unsafe|implicit|sizeof)\b</string>
        <key>name</key>
        <string>keyword.other.kit</string>
      </dict>
      <dict>
        <key>match</key>
        <string>\b(struct|enum|union|abstract|rule|rules|typedef|trait)\b</string>
        <key>name</key>
        <string>storage.type.kit</string>
      </dict>
      <dict>
        <key>name</key>
        <string>keyword.control.arrow.kit</string>
        <key>match</key>
        <string>=\&gt;</string>
      </dict>
      <dict>
        <key>name</key>
        <string>support.type.kit</string>
        <key>match</key>
        <string>\b[A-Z][a-zA-Z0-9_]*\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>variable.language.kit</string>
        <key>match</key>
        <string>\b(this|Self)\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>invalid.illegal.macro.kit</string>
        <key>match</key>
        <string>\$[A-Za-z0-9_]+\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>invalid.illegal.macro.kit</string>
        <key>begin</key>
        <string>\$\{</string>
        <key>end</key>
        <string>\}</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.decimal.kit</string>
        <key>match</key>
        <string>\b-?[0-9](\.[0-9+])?+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.octal.kit</string>
        <key>match</key>
        <string>\b-?0o[0-7]+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.hex.kit</string>
        <key>match</key>
        <string>\b-?0x[0-9a-f]+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.numeric.bin.kit</string>
        <key>match</key>
        <string>\b-?0b[01]+(_([ui](8|16|32|64)|f(32|64)))?\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>string.quoted.block.kit</string>
        <key>begin</key>
        <string>"""</string>
        <key>end</key>
        <string>"""</string>
      </dict>
      <dict>
        <key>name</key>
        <string>string.quoted.double.kit</string>
        <key>match</key>
        <string>"(\.|[^"])*"</string>
      </dict>
      <dict>
        <key>name</key>
        <string>string.quoted.single.kit</string>
        <key>match</key>
        <string>'(\.|[^\'])*'</string>
      </dict>
      <dict>
        <key>name</key>
        <string>comment.line.double-slash.kit</string>
        <key>match</key>
        <string>(//).*$
?</string>
      </dict>
      <dict>
        <key>name</key>
        <string>comment.block.kit</string>
        <key>begin</key>
        <string>/\*</string>
        <key>end</key>
        <string>\*/</string>
      </dict>
      <dict>
        <key>name</key>
        <string>constant.language.kit</string>
        <key>match</key>
        <string>\b(true|false|null)\b</string>
      </dict>
      <dict>
        <key>name</key>
        <string>markup.heading.metadata.kit</string>
        <key>begin</key>
        <string>\#\[</string>
        <key>end</key>
        <string>\]</string>
      </dict>
      <dict>
        <key>name</key>
        <string>variable.parameter.kit</string>
        <key>match</key>
        <string>\b[a-z_][A-Za-z0-9_]*\b</string>
      </dict>
    </array>
  </dict>
</plist>
