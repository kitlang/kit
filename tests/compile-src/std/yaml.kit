import kit.yaml;

function print_yaml(value: Yaml, indent: Size) {
    match value {
        YamlScalar(s) => {
            for i in 0 ... indent {
                printf("  ");
            }
            printf("%s\n", s);
        }
        YamlSequence(values) => {
            for value in values {
                for i in 0 ... indent {
                    printf("  ");
                }
                printf("-");
                match value {
                    YamlScalar(s) => {
                        printf(" %s\n", s);
                    }
                    default => {
                        printf("\n");
                        print_yaml(value, indent + 1);
                    }
                }
            }
        }
        YamlMap(pairs) => {
            for pair in pairs {
                for i in 0 ... indent {
                    printf("  ");
                }
                printf("%s:", pair[0]);
                match pair[1] {
                    YamlScalar(s) => {
                        printf(" %s\n", s);
                    }
                    default => {
                        printf("\n");
                        print_yaml(pair[1], indent + 1);
                    }
                }
            }
        }
    }
}

function main() {
    var result = Yaml.parse("""
str_key: hello!
bool_key: true
int_key: 1
null_key: null
nested_key1:
  nested_array:
  - 1
  - 2
array_key:
- 2
- false
- null
nested_key2:
  child_key: 4
array_of_objects:
- name: first_object
  value: 1
- name: second_object
  value: 2
""");

    var yaml = result.unwrap();
    print_yaml(yaml, 0);
    yaml.free();
}
