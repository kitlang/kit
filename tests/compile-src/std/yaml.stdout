str_key: hello!
bool_key: true
int_key: 1
null_key: null
nested_key1:
  nested_array:
    - 1
    - 2
array_key:
  - 2
  - false
  - null
nested_key2:
  child_key: 4
array_of_objects:
  -
    name: first_object
    value: 1
  -
    name: second_object
    value: 2
